Add better mechanism to detect if k8s caches are synced against k8s

aanm opened the pull request:
please read per commit message

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9400)
<!-- Reviewable:end -->

aanm committed:
pkg/lock: add stoppable waitgroup

As golang's WaitGroup does not offer the functionalities that we
require to detect if k8s caches were synchronized with the datapath.
This commit adds this initial library to which provides the ability to
signalize a caller when a set of go routines have finished their work.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
use StoppableWaitGroup to make sure k8s caches were synced with datapath

Signed-off-by: André Martins <andre@cilium.io>

brb commented on daemon/k8s_watcher.go:
```diff
@@ -461,7 +492,9 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
 					valid = true
+					swgSvcs.Add()
 					serSvcs.Enqueue(func() error {
+						defer swgSvcs.Done()
```
This won't work in my case (#9341). `d.addK8sServiceV1()` will call `k8s_service_cache` which might send a message (not in all cases) which is handled by `initK8sSubsystem()`. So, `swgSvcs.Done()` has to be called in `initK8sSubsystem()` after it has finished handling the event from the `k8s_service_cache`. See https://github.com/cilium/cilium/pull/9341/commits/10537f550b558007b979f7355f349ff6e4e00b39 for more details.

brb commented on daemon/k8s_watcher.go:
```diff
@@ -274,6 +282,8 @@ func (d *Daemon) blockWaitGroupToSyncResources(stop <-chan struct{}, informer ca
 		} else {
 			scopedLog.Debug("cache synced")
 		}
+		close(k8sCacheSynced)
```
I would add a method to `StoppableWaitGroup` to close the `noopAdd` (= `k8sCacheSynced`) instead. This would make the code more readable, as it wouldn't require a user to create and pass the channel to the constructor.

aanm committed:
k8s: wait until k8s services to be synchronized with datapath

With the help of StoppableWaitGroup we will be able to detect when the
services are correctly synchronized with the dapatah.

Signed-off-by: André Martins <andre@cilium.io>

brb commented on pkg/lock/stoppable_waitgroup.go:
```diff
@@ -0,0 +1,100 @@
+	default:
+		select {
+		case <-l.noopAdd:
+			a := atomic.AddUint64(l.i, ^uint64(0))
```
Shall we add a sanity check for the `< 0` condition?

aanm replied:
Sure, can you tell me one uint64 number that is `< 0`? :smile:

brb replied:
I meant reading the `l.i` value before decrementing it and checking whether it's equal to 0.

aanm replied:
@brb and panic in that case?

brb replied:
Yes, same as `sync.WaitGroup` does in this case.

brb commented on pkg/lock/stoppable_waitgroup_test.go:
```diff
@@ -0,0 +1,193 @@
+var _ = Suite(&StoppableWaitGroupSuite{})
+
+func (s *SemaphoredMutexSuite) TestAdd(c *C) {
+	l := &StoppableWaitGroup{
```
Nit: you can replace with `NewStoppableWaitGroup()`.

brb commented on daemon/k8s_watcher.go:
```diff
@@ -274,6 +281,8 @@ func (d *Daemon) blockWaitGroupToSyncResources(stop <-chan struct{}, informer ca
 		} else {
 			scopedLog.Debug("cache synced")
 		}
+		swg.Stop()
```
Nit: `swg.StopAndWait()`?

aanm replied:
I don't know it doesn't look good tbh

brb commented on daemon/state.go:
```diff
@@ -346,6 +346,14 @@ func (d *Daemon) initRestore(restoredEndpoints *endpointRestoreState) chan struc
 			if k8s.IsEnabled() {
+				// Also wait for all cluster mesh to be synchronized with the
+				// datapath before proceeding.
+				if option.Config.ClusterID != 0 {
```
Maybe more explicit option (e.g. as we have for k8s) to detect whether cluster-mesh is enabled?

aanm replied:
There's no such option

brb commented on daemon/state.go:
```diff
@@ -346,6 +346,14 @@ func (d *Daemon) initRestore(restoredEndpoints *endpointRestoreState) chan struc
+				// Also wait for all cluster mesh to be synchronized with the
+				// datapath before proceeding.
+				if option.Config.ClusterID != 0 {
+					err := d.clustermesh.ClustersSynced(context.Background())
```
Just double-checking, but do we know that at this point all remote clusters are known?

aanm replied:
We know that at this point all remote clusters are known because they are loaded from the directory. New files can show up after the cilium pod has started which means it's a new cluster and wasn't running before cilium restart.

brb commented on daemon/state.go:
```diff
@@ -346,6 +346,14 @@ func (d *Daemon) initRestore(restoredEndpoints *endpointRestoreState) chan struc
+				if option.Config.ClusterID != 0 {
+					err := d.clustermesh.ClustersSynced(context.Background())
+					if err != nil {
+						log.Fatal(err)
```
Perhaps add more context to the log message?

brb reviewed (changes requested):
In general LGTM, a couple of nits.

I'd personally avoid passing the `StoppableWaitGroup` instance to the `ServiceCache` methods even if it would result in more lines of code, and would call `Add()` and `Done()` only from `daemon/k8s_watcher.go`.

brb reviewed (changes requested):
I think the following race condition still exists which can result in a deadlock, if there are no more calls to `l.Done()`:
```
l.i = 1, l.noopDone and l.noopAdd are not closed.

# thread 1                                  # thread 2
-----------------------------------------------------------------------------------
l.Stop() {                                  l.Done() {
	l.doneOnce.Do(func() {                  select {
                                                  case <-l.noopDone:                        
                                                  	return
                                                  default:
                                                  	select {
                                                  	case <-l.noopAdd:
                                                  		a := atomic.AddInt64(l.i, -1)
                                                  		if a <= 0 {
                                                  			l.wgOnce.Do(func() {
                                                  				close(l.noopDone)
                                                  			})
                                                  		}
                                                  	default:    
                           
		close(l.noopAdd)                      	
		if atomic.LoadInt64(l.i) <= 0 {       	
			l.wgOnce.Do(func() {              	
				close(l.noopDone)             	
			})                                	
		}                                     	
	})                                        	
}                                             	

                                              			atomic.AddInt64(l.i, -1)
                                              		}
                                              	}
											  }
```

Also, I think you should explicitly mention in the docs that the stoppable WG tolerates negative counters, and that `Wait()` unblocks once the counter becomes a non-positive/

brb reviewed (changes requested):
LGTM, except the fact that `k8s_service_cache.go` should know about a stoppable WG which complicates the code of the cache.

Can you add an unit test where multiple goroutines do `Add()` and `Done()` just to test a concurrent access of the library (in the end it will be used by multiple goroutines)?
